### DISCUSSION TITLE
Need help in understanding what checks the pyHanko does during validation

### DISCUSSION THREAD

**Discussion by Vishwamithra37** [MatthiasValvekens/pyHanko#119] [Q&A]
I understand that as you said, you use asn1crypto to read and validate the signatures. But can you tell me exactly what checks happen when I try to validate through pyHanko? In your own words as the maintainer and developer?

**Comment by MatthiasValvekens** [comment]
Let me preface this by saying that PDF signature validation is a lot more complicated than it appears, and pyHanko is not yet where I want it to be, esp. when it comes to long-term trust validation.

Here's an overview of what has to be checked (not necessarily in this order):

 - Basic cryptographic integrity checks: compare the document digest to the message digest attribute in the signature, and validate the actual cryptographic signature itself (over the signed attributes) using the signer's public key. This is the trivial part.
 - Trust validation: validate the binding between the signer's public key and their identity, AKA certificate validation. This is a multi-step process: find candidate paths from the signer's certificate to one or more trust roots, validate these paths (basic principles: RFC 5280), check revocation status. FWIW, this process is a lot more complicated for signatures that have to outlive their certificates, and pyHanko's validation logic doesn't deal with those all too sensibly right now.
 - Establishing the time of signing: closely related to trust validation and with similar complexity.
 - PDF incremental update validation: for documents with multiple signatures and/or with changes applied after signing (by appending data to the end of the document), it's important to check whether those post-signing changes are legitimate or not. This process is not standardised at all.
- Seed value constraint enforcement: another PDF-specific thing that you probably won't have to worry about.

There's more info [in the docs](https://pyhanko.readthedocs.io/en/latest/cli-guide/validation.html).

**Reply by Vishwamithra37** [reply]
Okay, so, I have a few questions:

1. If a PDF has form field, where a user can fill their name details etc. Is there a facility for pyHanko to selectively ignore those fields and make a digest out of the rest of the PDF?Or does it always involve whole file message digest/hash.
2. I am a little confused about the trust validation, Should the root cert.pem match the one in the certificate to get trust validated?
3. Isn't it as simple as putting in the time stamp? in case of time of signing. 
4. I reall don't understand what seed value is but would love to know more about it :).

In my application, I am just storing the details after a successful signing in mongoDB. That is:

1. I do multiple random hashes of the file and store it. 
2. I am trying to get the public-key out of the cert object using asn1crypto but I am unable to do so. Help plzzz :). So that I can test and match if the public key is the correct pair of the private key.
Is this secure enough? any thoughts would be realllllly helpful thanks for your time .

**Reply by MatthiasValvekens** [reply]
> If a PDF has form field, where a user can fill their name details etc. Is there a facility for pyHanko to selectively ignore those fields and make a digest out of the rest of the PDF?Or does it always involve whole file message digest/hash.

No. There's no sane way to selectively exclude form fields from signature coverage. What you _can_ do is use field locks to flag to the validator that certain form fields may be changed after signing (through incremental updates), while others must remain untouched. But unless you really have no choice, I strongly suggest not to go down that path. Most PDF signature validators don't even try to implement any of this stuff properly (it's also extremely poorly specified), and you should strive to avoid situations where your signatures are subject to ambiguity.

> I am a little confused about the trust validation, Should the root cert.pem match the one in the certificate to get trust validated?

You're not alone, and the answer is no. The signer's certificate should have a valid _path_ to at least one certificate authority that is considered a trust root by the validator. In practice, there's often only one candidate path, and the signer and validator both trust the root of that path. Have a look at [the Wikipedia page about CAs](https://en.wikipedia.org/wiki/Certificate_authority) for more info about how CAs work at the conceptual level, and read [RFC 5280](https://tools.ietf.org/html/rfc5280) if you want to know how path validation _actually_ works.

> Isn't it as simple as putting in the time stamp? in case of time of signing.

Not if you want the timestamp to be cryptographically enforceable. A trusted timestamp can be used to prove that a signature/document/piece of data/... existed at a given time. Merely embedding the time of signing into the signature doesn't achieve that. Without trusted timestamps, it's impossible to prevent an attacker with a stolen key from backdating signatures to within the validity period of the certificate.

By extension (and this is really the more serious problem!), it would follow that certificate expiry/revocation invalidates all signatures made based on the key in that certificate. For document signing, that's _very, very bad_: after all, you wouldn't want someone to be able to get out of a contract by "conveniently" losing their key.

> I reall don't understand what seed value is but would love to know more about it :).

Long story short, they're PDFs way of providing guidance to an application signing a PDF form. It was a thing in Adobe LiveCycle in the days of yore (that product is dead now), and made it into the ISO standard(s) for PDF, but it's very poorly supported in the wild. I put basic seed value support into pyHanko because it was pretty easy, but in practice nobody cares and you shouldn't rely on other software supporting it ;).


> I do multiple random hashes of the file and store it.

Sorry, I don't understand. What are "random hashes", and why would you ever randomise hash inputs for integrity checks? Seems counterproductive ;).

>  I am trying to get the public-key out of the cert object using asn1crypto but I am unable to do so. Help plzzz :). So that I can test and match if the public key is the correct pair of the private key.

Well, I could explain how to perform that extraction (pyHanko can extract the signer cert for you as an `asn1crypto` object) but again I don't understand why you're trying to do that check. A validator shouldn't have access to the private key either way, so by definition this can't affect the validity of the signature.

>  Is this secure enough? any thoughts would be realllllly helpful thanks for your time .

...that's impossible for me to judge. What's your threat model? What exactly are you trying to protect here? Ask yourself these questions, and then hire a security consultant to tell you what that means for your app design, because design issues of that type are light-years beyond the scope of this discussion board ;).

**Reply by Vishwamithra37** [reply]
Hey, thanks for the reply, I am still unable to see document digest, which is nothing but hash, I presume.
Here is what exactly I am trying to do:

- I have 1 document signed with two signatures.
- I am trying to compute the digest using 
`      signature_objec = pdf_reader.embedded_signatures[1]
        print(signature_objec.compute_digest().hex())`

- However I am getting different hex() values for each signature. 
- I **Understand this is the signature digest and not the document digest. How do I get the document digest?**
Like I want to verify that the document has not been modified for the two signatures. 



Random hashses in my sense mean, taking out a random length of bytes and hashing them instead of the whole hash. This way even if an attacker somehow manages to produce the same hash but with different contents, the validation would find out. 

sorry if I am messing up my whole understanding.

**Reply by MatthiasValvekens** [reply]
Sorry for the delay in replying.

> However I am getting different hex() values for each signature.

That's entirely normal. PDF signatures are computed serially, so signing the same doc multiple times will produce different document digests each time.

> Random hashses in my sense mean, taking out a random length of bytes and hashing them instead of the whole hash. This way even if an attacker somehow manages to produce the same hash but with different contents, the validation would find out.

Don't do that. Hash functions weren't designed for this kind of usage, and I'd be willing to bet that this construction is actually _less_ secure in practice compared to just hashing the whole thing :).

**Reply by slice-saurabha** [reply]
@MatthiasValvekens @Vishwamithra37 I'm also getting different digest values every time I sign the same PDF file, with the same pfx certificate. Why is it happening? The signed PDF file is stored with a different name.

Also, how can I get the doc digest from inside the signature. I want to compare it manually with the doc digest before signing to verify the content has not modified.

**Reply by Vishwamithra37** [reply]
I simply took the file hash and used some other library to remove the sign and verify the hash. 

As it stands it's already implemented by multitude of companies in India and my  implementation of the same is done too :) at https://galam.in

**Reply by slice-saurabha** [reply]
@MatthiasValvekens Is there any way to do it in Pyhanko.

My usecase: I'm signing a PDF with a single signer. I want to verify that the SHA256 hash for the unsigned PDF file matches with the hash returned from  signature_objec.compute_digest().hex().

Thanks in advance.

**Reply by slice-saurabha** [reply]
@Vishwamithra37 Which library did you use to remove signature, and did you figure out why compute_digest() returns different hash everytime?

**Reply by Vishwamithra37** [reply]
https://pypi.org/project/pdfrw2/https://pypi.org/project/pdfrw2/
